### DISCUSSION TITLE
Dbus-serialbattery switching between USB ports

### DISCUSSION THREAD

**Discussion by hansgeorgk** [Louisvdw/dbus-serialbattery#265] [Q&A]
Hi,

after installation of v01.3 it seems like in my driver is switching between USB port 1 and 3

drwx------    2 root     root          4096 Oct 15 06:24 dbus-serialbattery.ttyUSB1
drwx------    2 root     root          4096 Oct 15 06:27 dbus-serialbattery.ttyUSB3

but does not work properly on 3 and then is switching back to 1
Every few minutes the connection is lost in Remote console and now values are shown for a few seconds ...

any idea how to fix this?
cheers
HG

**Comment by Louisvdw** [comment]
The only way ports can switch would be if there is some new USB device that is plugged in. 
Could it be that there is a USB that is not making good contact? Remove all USB connections and plug them back one at a time to see which one is not behaving. Give a few minutes between each one that you plug back in. Your GX device need to cycle through all the drivers to start on the new port.

**Comment by hansgeorgk** [comment]
re-installed Venus OS on my Raspi - latest Release v2.92.
installed one device after the other - still having issues with serialbattery. Unplugged all devices (BMV, MTTP and Multiplus)
Serialbattery is not switching between ports anymore but is showing Protocol errors :-(
Besides the "ERROR: No reply", which is coming every second, the log file is showing "OSError: [Errno 71] Protocol error" every other minute ...


@400000006352fd4e350952ec ERROR:SerialBattery:>>> ERROR: No reply - returning
@400000006352fd5035946c4c ERROR:SerialBattery:>>> ERROR: No reply - returning
@400000006352fd5226bc7354 Traceback (most recent call last):
@400000006352fd5226bc8ac4   File "/opt/victronenergy/dbus-serialbattery/dbushelper.py", line 163, in publish_battery
@400000006352fd5226bc9e4c     success = self.battery.refresh_data()
@400000006352fd5226bcaa04   File "/opt/victronenergy/dbus-serialbattery/daly.py", line 60, in refresh_data
@400000006352fd5226bcbd8c     ser = open_serial_port(self.port, self.baud_rate)
@400000006352fd5226bcc944   File "/opt/victronenergy/dbus-serialbattery/utils.py", line 98, in open_serial_port
@400000006352fd5226be6754     ser = serial.Serial(port, baudrate=baud, timeout=0.1)
@400000006352fd5226be76f4   File "/usr/lib/python3.8/site-packages/serial/serialutil.py", line 240, in __init__
@400000006352fd5226be8a7c     self.open()
@400000006352fd5226be924c   File "/usr/lib/python3.8/site-packages/serial/serialposix.py", line 286, in open
@400000006352fd5226bea5d4     self._update_dtr_state()
@400000006352fd5226bf0394   File "/usr/lib/python3.8/site-packages/serial/serialposix.py", line 634, in _update_dtr_state
@400000006352fd5226bf1b04     fcntl.ioctl(self.fd, TIOCMBIS, TIOCM_DTR_str)
@400000006352fd5226bf26bc OSError: [Errno 71] Protocol error
@400000006352fd63051c384c INFO:SerialBattery:Starting dbus-serialbattery
@400000006352fd6305272cfc INFO:SerialBattery:dbus-serialbattery v0.13
@400000006352fd630548e1e4 INFO:SerialBattery:Testing LltJbd
@400000006352fd63154ab8c4 ERROR:SerialBattery:>>> ERROR: No reply - returning
@400000006352fd63155dbfdc INFO:SerialBattery:Testing Ant
@400000006352fd6325937414 ERROR:SerialBattery:>>> ERROR: No reply - returning
@400000006352fd6325a62154 ERROR:SerialBattery:>>> ERROR: Incorrect Data
@400000006352fd6325af740c INFO:SerialBattery:Testing Daly
@400000006352fd6329277fe4 INFO:SerialBattery:DalyBMS 4 cells
@400000006352fd63293b59ec INFO:SerialBattery:Connection established to Daly
@400000006352fd6329454ccc INFO:SerialBattery:Battery connected to dbus from /dev/ttyUSB3
@400000006352fd63294fb4dc INFO:SerialBattery:=== Settings ===
@400000006352fd632958ccfc INFO:SerialBattery:> Connection voltage NoneV | current NoneA | SOC None%
@400000006352fd6329626604 INFO:SerialBattery:> Cell count 4 | cells populated 0
@400000006352fd63296aaf1c INFO:SerialBattery:> CCL Charge NoneA | DCL Discharge NoneA
@400000006352fd6329768274 INFO:SerialBattery:> MIN_CELL_VOLTAGE 2.7V | MAX_CELL_VOLTAGE 3.55V
@400000006352fd632c2ffd24 INFO:SerialBattery:DeviceInstance = 1
@400000006352fd632c3c9f84 INFO:SerialBattery:com.victronenergy.battery.ttyUSB3
@400000006352fd6702899ebc ERROR:SerialBattery:>>> ERROR: No reply - returning
@400000006352fd6902361404 ERROR:SerialBattery:>>> ERROR: No reply - returning
@400000006352fd6b02c20ff4 ERROR:SerialBattery:>>> ERROR: No reply - returning
@400000006352fd6d031d8dd4 ERROR:SerialBattery:>>> ERROR: No reply - returning
@400000006352fd6f02c870c4 ERROR:SerialBattery:>>> ERROR: No reply - returning
@400000006352fd71031fb884 ERROR:SerialBattery:>>> ERROR: No reply - returning

seems like there is also some vedirect service trying to access same USB
@400000006353014133f8e0d4 -------- dumping backlog -------
@400000006353014133f8ec8c 000.023 INF.task: using ttyUSB3 as unique service part
@400000006353014133f8f844 000.024 INF.VE.Direct: Setting status to VdUnknown
@400000006353014133f907e4 000.032 INF.settings: connected to settings on dbus (com.victronenergy.settings)
@400000006353014133f91784 002.999 ERR.values: timeout connecting device
@400000006353014133f9233c 002.999 INF.task: done in 002.999
@400000006353015e2c2308bc *** starting vedirect-interface ***
@400000006353015e2d85e864 vedirect-dbus 3.70
@40000000635301612d36ab3c
@40000000635301612d36bec4 -------- dumping backlog -------
@40000000635301612d36ca7c 000.022 INF.task: using ttyUSB3 as unique service part
@40000000635301612d36d634 000.023 INF.VE.Direct: Setting status to VdUnknown
@40000000635301612d36e5d4 000.032 INF.settings: connected to settings on dbus (com.victronenergy.settings)
@40000000635301612d36f574 002.995 ERR.values: timeout connecting device
@40000000635301612d37012c 002.995 INF.task: done in 002.995
root@raspberrypi2:/data/log/vedirect.ttyUSB3#


in messages I see
Oct 21 20:34:02 raspberrypi2 user.err kernel: [ 2381.010921] usb 1-1.1.3.1.1: failed to send control message: -71
Oct 21 20:34:38 raspberrypi2 user.err kernel: [ 2417.081939] usb 1-1.1.3.1.1: failed to send control message: -71
Oct 21 20:34:55 raspberrypi2 user.err kernel: [ 2433.311893] usb 1-1.1.3.1.1: failed to send control message: -71

**Reply by Louisvdw** [reply]
change the polling interval from 1000 to 1500 for your Daly in dbus-serialbattery.py and then restart your GX.
See if this solves your error.

**Comment by hansgeorgk** [comment]
cannot find polling interval ind bus-serialbattery.py ...
I can find it multiple time at various places 
root@raspberrypi2:/data/etc/dbus-serialbattery# grep interval *
battery.py:        self.poll_interval = 1000
daly.py:        self.poll_interval = 1000
dbus-serialbattery.py:    gobject.timeout_add(battery.poll_interval, lambda: poll_battery(mainloop))
dbushelper.py:        # This is called every battery.poll_interval milli second as set up per battery type to read and update the data
minimalmodbus.py:        for example to fulfill the inter-message time interval or to give
mnb.py:        self.poll_interval = None
mnb.py:        self.poll_interval = 1000 #scan repeat time, ms
sinowealth.py:        self.poll_interval = 2000
util_max17853.py:        self.poll_interval = 3000
util_max17853.py:        self.poll_interval = 1000
utils.py:# TIME TO SOC settings [Valid values 0-100, but I don't recommend more that 20 intervals]

where shall I change it?

**Reply by Louisvdw** [reply]
sorry. I meant in the daly driver. It would be this one
> daly.py: self.poll_interval = 1000

**Comment by hansgeorgk** [comment]
tried 1500 and 2000  - no luck :-(



@4000000063557b02280fe9ac ERROR:SerialBattery:>>> ERROR: No reply - returning
@4000000063557b062767c2cc ERROR:SerialBattery:>>> ERROR: No reply - returning
@4000000063557b0a28567c14 ERROR:SerialBattery:>>> ERROR: No reply - returning
@4000000063557b0e2926d01c ERROR:SerialBattery:>>> ERROR: No reply - returning

**Reply by Louisvdw** [reply]
Can you check in your phone app if the BMS has not gone to sleep. And then set the sleep timeout to the largest value it can use.

**Comment by hansgeorgk** [comment]
the sleep timeout is set to default of 3600s ...
Don't see any other time settings in Daly BMS Phone App ...
Are there more available using the PC Software?

**Reply by Louisvdw** [reply]
Make that sleep timeout as large as you can.

**Comment by hansgeorgk** [comment]
In the meantime I switched to an EasySolar II GX :-). 
Ports are not switching any more - serialbattdry is running continuously on USB1 now.

I tried 99999, it did not accept the number. I finally found 65535 working as big value for sleep time in Daly settings.

I am still getting an error. after reboot it is connecting, but the error continues to show up :-(

In dashboard serialbattery is going to idle mode every 3-5 seconds

@4000000063601cc6063e5da4 ERROR:SerialBattery:>>> ERROR: No reply - returning
@4000000063601cc806c05ee4 ERROR:SerialBattery:>>> ERROR: No reply - returning
@4000000063601cca06ac753c ERROR:SerialBattery:>>> ERROR: No reply - returning
@4000000063601cdd32178a2c *** CCGX booted (0) ***
@4000000063601ceb27023e84 INFO:SerialBattery:Starting dbus-serialbattery
@4000000063601ceb2713901c INFO:SerialBattery:dbus-serialbattery v0.13
@4000000063601ceb274108bc INFO:SerialBattery:Testing LltJbd
@4000000063601ceb38d25554 ERROR:SerialBattery:>>> ERROR: No reply - returning
@4000000063601ceb39436a8c INFO:SerialBattery:Testing Ant
@4000000063601cec0e76434c ERROR:SerialBattery:>>> ERROR: No reply - returning
@4000000063601cec0ee1900c ERROR:SerialBattery:>>> ERROR: Incorrect Data
@4000000063601cec0ef2417c INFO:SerialBattery:Testing Daly
@4000000063601cec120bd1d4 INFO:SerialBattery:DalyBMS 8 cells
@4000000063601cec129d099c INFO:SerialBattery:Connection established to Daly
@4000000063601cec12a67b94 INFO:SerialBattery:Battery connected to dbus from /dev/ttyUSB1
@4000000063601cec12acc10c INFO:SerialBattery:=== Settings ===
@4000000063601cec12b3a6ac INFO:SerialBattery:> Connection voltage NoneV | current NoneA | SOC None%
@4000000063601cec12b9d4b4 INFO:SerialBattery:> Cell count 8 | cells populated 0
@4000000063601cec12c1bc24 INFO:SerialBattery:> CCL Charge NoneA | DCL Discharge NoneA
@4000000063601cec12c8578c INFO:SerialBattery:> MIN_CELL_VOLTAGE 2.6V | MAX_CELL_VOLTAGE 3.55V
@4000000063601cec1550719c INFO:SerialBattery:DeviceInstance = 1
@4000000063601cec155a31b4 INFO:SerialBattery:com.victronenergy.battery.ttyUSB1
@4000000063601cef29244b94 ERROR:SerialBattery:>>> ERROR: No reply - returning
@4000000063601cf127fcabe4 ERROR:SerialBattery:>>> ERROR: No reply - returning
@4000000063601cf327b48724 ERROR:SerialBattery:>>> ERROR: No reply - returning
@4000000063601cf527e7242c ERROR:SerialBattery:>>> ERROR: No reply - returning

**Reply by Louisvdw** [reply]
If the battery continue after that error then it is fine. 
Sometimes the communication between the GX and the BMS get interference. Then that reply will be ignored. The driver and BMS should recover and continue again.
If you see the no reply and it does not continue then that is a issue we can look at.
